File: bug_fab/schemas.py
"""Pydantic v2 schemas for the Bug-Fab wire protocol.

These models define the payload shape for:
- Submission intake (`BugReportCreate` + `BugReportContext`)
- Status updates (`BugReportStatusUpdate`)
- Read responses (`BugReportSummary`, `BugReportDetail`, `BugReportListResponse`)
- Lifecycle audit entries (`LifecycleEvent`)

Severity and Status are real string enums (subclasses of `str` + `enum.Enum`)
so JSON round-trips and adapter validation share one source of truth.
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class Severity(str, Enum):
    """Locked severity vocabulary — adapters MUST reject other values with 422.

    WHY str-Enum (not StrEnum): Python 3.10 support is required, and StrEnum
    was added in 3.11. Subclassing both `str` and `Enum` produces equivalent
    JSON behavior on the older interpreter.
    """

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class Status(str, Enum):
    """Locked status vocabulary for the lifecycle workflow.

    WHY four values: matches every audited prior implementation. Deprecated
    values (e.g., legacy `resolved`) MUST still parse on read per protocol;
    write-side validation is strict.
    """

    OPEN = "open"
    INVESTIGATING = "investigating"
    FIXED = "fixed"
    CLOSED = "closed"


class Reporter(BaseModel):
    """Optional submitter identity attached to a bug report.

    All fields are opaque strings, capped at 256 characters per the
    2026-04-28 spec-gap decisions (per `docs/decisions.md`). The protocol
    does not validate the format of these values — consumer user IDs vary
    widely (UUIDs, emails, integers-as-strings, SSO subjects).

    Adapters MAY override the client-supplied values with server-derived
    identity when they have their own auth context.
    """

    model_config = ConfigDict(extra="ignore")

    name: str = Field(default="", max_length=256)
    email: str = Field(default="", max_length=256)
    user_id: str = Field(default="", max_length=256)


class BugReportContext(BaseModel):
    """Auto-captured browser context attached to every submission.

    `source_mapping` is consumer-supplied opaque metadata; the server does
    NOT compute URL→file mappings (that responsibility moved to the client
    or to a consumer-provided resolver hook in v0.2+).

    `user_agent` here is the *client-reported* value, kept for diagnostics.
    The adapter captures the request-header User-Agent separately as the
    source of truth (`server_user_agent` on `BugReportDetail`).
    """

    model_config = ConfigDict(extra="allow")

    url: str = ""
    module: str = ""
    user_agent: str = ""
    viewport_width: int = 0
    viewport_height: int = 0
    console_errors: list[dict[str, Any]] = Field(default_factory=list)
    network_log: list[dict[str, Any]] = Field(default_factory=list)
    source_mapping: dict[str, Any] = Field(default_factory=dict)
    app_version: str = ""
    environment: str = ""


class BugReportCreate(BaseModel):
    """Submission payload sent as a JSON string in the multipart `metadata` field.

    `report_type` is a `Literal` (not enum) because the protocol freezes the
    two values and there is no read-side deprecation concern.

    `protocol_version` is required and must equal `"0.1"`. Future versions
    bump this. Adapters MUST reject unknown values with `400 unsupported_protocol_version`.

    `client_ts` is a client-side ISO-8601 timestamp captured when the user
    pressed Submit. Diagnostic only — the server's `received_at` /
    `created_at` is the authoritative timeline.
    """

    model_config = ConfigDict(extra="ignore")

    protocol_version: Literal["0.1"]
    title: str = Field(min_length=1, max_length=200)
    client_ts: str = Field(min_length=1)
    report_type: Literal["bug", "feature_request"] = "bug"
    description: str = ""
    expected_behavior: str = ""
    severity: Severity = Severity.MEDIUM
    tags: list[str] = Field(default_factory=list)
    reporter: Reporter = Field(default_factory=Reporter)
    context: BugReportContext = Field(default_factory=BugReportContext)


class BugReportStatusUpdate(BaseModel):
    """Body of `PUT /reports/{id}/status` — strict severity-style validation."""

    model_config = ConfigDict(extra="ignore")

    status: Status
    fix_commit: str = ""
    fix_description: str = ""


class LifecycleEvent(BaseModel):
    """One entry in a report's lifecycle audit log.

    Field names lock to `action / by / at` per audit IF16 — the prior-art
    template/service drift (`status / changed_by / timestamp`) is the
    cautionary tale that motivated the lock.
    """

    model_config = ConfigDict(extra="allow")

    action: str
    by: str = ""
    at: str
    fix_commit: str = ""
    fix_description: str = ""


class BugReportSummary(BaseModel):
    """Compact representation used by list views and the index.

    `module`, `github_issue_url`, and `has_screenshot` are denormalized
    onto the summary so listings render without re-reading every detail
    file or row.
    """

    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    report_type: str = "bug"
    severity: str = Severity.MEDIUM.value
    status: str = Status.OPEN.value
    module: str = ""
    created_at: str
    has_screenshot: bool = True
    github_issue_url: str | None = None


class BugReportDetail(BugReportSummary):
    """Full report payload returned by detail/status endpoints.

    Extends `BugReportSummary` with everything the viewer detail panel
    needs: free-text fields, tags, the captured context, lifecycle log,
    and the dual user-agent fields (server-captured + client-reported).

    `client_reported_user_agent` mirrors `context.user_agent` but is
    surfaced at the top level so consumers and viewers do not have to
    reach into the nested context to compare them.
    """

    description: str = ""
    expected_behavior: str = ""
    tags: list[str] = Field(default_factory=list)
    reporter: Reporter = Field(default_factory=Reporter)
    context: BugReportContext = Field(default_factory=BugReportContext)
    lifecycle: list[LifecycleEvent] = Field(default_factory=list)
    server_user_agent: str = ""
    client_reported_user_agent: str = ""
    environment: str = ""
    client_ts: str = ""
    protocol_version: str = "0.1"
    updated_at: str = ""
    github_issue_number: int | None = None


class BugReportListResponse(BaseModel):
    """Pagination envelope for `GET /reports`."""

    model_config = ConfigDict(extra="ignore")

    items: list[BugReportSummary]
    total: int
    page: int = 1
    page_size: int = 20

File: adapters/rust-axum/bugfab/src/schemas.rs
//! Wire-protocol schemas for Bug-Fab v0.1.
//!
//! Mirrors `bug_fab/schemas.py` from the Python reference. Enums use
//! `#[serde(rename_all = "lowercase")]` so the on-wire vocabulary matches
//! the protocol exactly. Deserialization is strict — unknown enum values
//! produce a serde error which the HTTP layer converts into 422.
//!
//! The Reporter / Context / Detail structs are kept extra-tolerant
//! (preserving unknown keys via `serde_json::Map<String, Value>` on
//! `BugReportContext.extra`) so consumer-specific diagnostic fields
//! survive round-trip per PROTOCOL.md.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

/// Locked severity vocabulary. Adapters MUST reject other values with 422.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Low,
    Medium,
    High,
    Critical,
}

impl Default for Severity {
    fn default() -> Self {
        Severity::Medium
    }
}

impl Severity {
    pub fn as_wire(&self) -> &'static str {
        match self {
            Severity::Low => "low",
            Severity::Medium => "medium",
            Severity::High => "high",
            Severity::Critical => "critical",
        }
    }
}

/// Locked status vocabulary for the lifecycle workflow.
///
/// Note: per PROTOCOL.md §"Deprecated-values rule", adapters MUST accept
/// deprecated values on read paths. The viewer's storage layer therefore
/// reads `status` as a raw `String` (the schemas surface it as such on
/// `BugReportDetail`/`BugReportSummary`) and only this `Status` enum is
/// used for write-side validation.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Status {
    Open,
    Investigating,
    Fixed,
    Closed,
}

impl Status {
    pub fn as_wire(&self) -> &'static str {
        match self {
            Status::Open => "open",
            Status::Investigating => "investigating",
            Status::Fixed => "fixed",
            Status::Closed => "closed",
        }
    }
}

/// Report type — frozen for v0.1.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReportType {
    Bug,
    FeatureRequest,
}

impl Default for ReportType {
    fn default() -> Self {
        ReportType::Bug
    }
}

impl ReportType {
    pub fn as_wire(&self) -> &'static str {
        match self {
            ReportType::Bug => "bug",
            ReportType::FeatureRequest => "feature_request",
        }
    }
}

/// Protocol version literal — only `"0.1"` is accepted on write.
pub const PROTOCOL_VERSION: &str = "0.1";

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Reporter {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub email: String,
    #[serde(default)]
    pub user_id: String,
}

impl Reporter {
    /// Per PROTOCOL.md — sub-fields are opaque strings, capped at 256 chars.
    pub fn validate(&self) -> Result<(), String> {
        for (field, value) in [
            ("reporter.name", &self.name),
            ("reporter.email", &self.email),
            ("reporter.user_id", &self.user_id),
        ] {
            if value.len() > 256 {
                return Err(format!("{} exceeds 256 character cap", field));
            }
        }
        Ok(())
    }
}

/// Auto-captured browser context. Extra keys are preserved via `extra`.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct BugReportContext {
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub module: String,
    #[serde(default)]
    pub user_agent: String,
    #[serde(default)]
    pub viewport_width: u32,
    #[serde(default)]
    pub viewport_height: u32,
    #[serde(default)]
    pub console_errors: Vec<Value>,
    #[serde(default)]
    pub network_log: Vec<Value>,
    #[serde(default)]
    pub source_mapping: Map<String, Value>,
    #[serde(default)]
    pub app_version: String,
    #[serde(default)]
    pub environment: String,

    /// Forward-additive bag — preserves consumer-specific diagnostic keys.
    #[serde(flatten)]
    pub extra: Map<String, Value>,
}

/// Submission payload (the JSON string in the multipart `metadata` field).
#[derive(Debug, Clone, Deserialize)]
pub struct BugReportCreate {
    pub protocol_version: String,
    pub title: String,
    pub client_ts: String,
    #[serde(default)]
    pub report_type: ReportType,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub expected_behavior: String,
    #[serde(default)]
    pub severity: Severity,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub reporter: Reporter,
    #[serde(default)]
    pub context: BugReportContext,
}

impl BugReportCreate {
    /// Per-field validation that serde alone can't express.
    ///
    /// Returns the human-readable detail string for a 422 response.
    pub fn validate(&self) -> Result<(), String> {
        if self.protocol_version != PROTOCOL_VERSION {
            return Err(format!(
                "unsupported protocol_version: {:?}",
                self.protocol_version
            ));
        }
        let title_len = self.title.chars().count();
        if title_len == 0 || title_len > 200 {
            return Err("title must be 1..=200 characters".to_string());
        }
        if self.client_ts.is_empty() {
            return Err("client_ts is required".to_string());
        }
        self.reporter.validate()?;
        Ok(())
    }
}

/// Body of `PUT /reports/{id}/status` — strict enum validation.
#[derive(Debug, Clone, Deserialize)]
pub struct BugReportStatusUpdate {
    pub status: Status,
    #[serde(default)]
    pub fix_commit: String,
    #[serde(default)]
    pub fix_description: String,
}

/// One entry in a report's lifecycle audit log.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifecycleEvent {
    pub action: String,
    #[serde(default)]
    pub by: String,
    pub at: String,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub fix_commit: String,
    #[serde(default)]
    pub fix_description: String,
}

/// Summary representation used for the list endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BugReportSummary {
    pub id: String,
    #[serde(default)]
    pub title: String,
    #[serde(default = "default_report_type")]
    pub report_type: String,
    #[serde(default = "default_severity")]
    pub severity: String,
    #[serde(default = "default_status")]
    pub status: String,
    #[serde(default)]
    pub module: String,
    pub created_at: String,
    #[serde(default = "default_true")]
    pub has_screenshot: bool,
    #[serde(default)]
    pub github_issue_url: Option<String>,
}

fn default_report_type() -> String {
    "bug".to_string()
}
fn default_severity() -> String {
    "medium".to_string()
}
fn default_status() -> String {
    "open".to_string()
}
fn default_true() -> bool {
    true
}

/// Full detail payload returned by detail / status endpoints.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BugReportDetail {
    pub id: String,
    pub title: String,
    pub report_type: String,
    pub severity: String,
    pub status: String,
    #[serde(default)]
    pub module: String,
    pub created_at: String,
    pub has_screenshot: bool,
    #[serde(default)]
    pub github_issue_url: Option<String>,

    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub expected_behavior: String,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub reporter: Reporter,
    #[serde(default)]
    pub context: BugReportContext,
    #[serde(default)]
    pub lifecycle: Vec<LifecycleEvent>,
    #[serde(default)]
    pub server_user_agent: String,
    #[serde(default)]
    pub client_reported_user_agent: String,
    #[serde(default)]
    pub environment: String,
    #[serde(default)]
    pub client_ts: String,
    #[serde(default = "default_protocol_version")]
    pub protocol_version: String,
    #[serde(default)]
    pub updated_at: String,
    #[serde(default)]
    pub github_issue_number: Option<i64>,
}

fn default_protocol_version() -> String {
    PROTOCOL_VERSION.to_string()
}

/// Pagination envelope for `GET /reports`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BugReportListResponse {
    pub items: Vec<BugReportSummary>,
    pub total: u64,
    pub page: u32,
    pub page_size: u32,
    pub stats: std::collections::BTreeMap<String, u64>,
}

/// Minimal `201 Created` body for `POST /bug-reports`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BugReportIntakeResponse {
    pub id: String,
    pub received_at: String,
    pub stored_at: String,
    pub github_issue_url: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn severity_invalid_value_rejected() {
        let payload = r#"{"protocol_version":"0.1","title":"x","client_ts":"now","severity":"urgent"}"#;
        let err = serde_json::from_str::<BugReportCreate>(payload).unwrap_err();
        let msg = err.to_string();
        assert!(msg.contains("urgent") || msg.contains("variant"), "{msg}");
    }

    #[test]
    fn severity_default_is_medium() {
        let payload = r#"{"protocol_version":"0.1","title":"x","client_ts":"now"}"#;
        let parsed: BugReportCreate = serde_json::from_str(payload).unwrap();
        assert_eq!(parsed.severity.as_wire(), "medium");
    }

    #[test]
    fn protocol_version_validate_rejects_unknown() {
        let payload = r#"{"protocol_version":"0.2","title":"x","client_ts":"now"}"#;
        let parsed: BugReportCreate = serde_json::from_str(payload).unwrap();
        let err = parsed.validate().unwrap_err();
        assert!(err.contains("protocol_version"), "{err}");
    }

    #[test]
    fn context_preserves_extra_keys() {
        let payload = r#"{"url":"x","custom_field":"keep me"}"#;
        let parsed: BugReportContext = serde_json::from_str(payload).unwrap();
        assert_eq!(parsed.extra.get("custom_field").unwrap(), "keep me");
    }
}
